### PR TITLE
[PTQ][OV] FP8 implementation

### PR DESCRIPTION
**Pull request by KodiaqQ**
### Changes

- Added FP8 implementation
- Added `Mode` parameter

### Reason for changes

- New FP8 implementation

### Related tickets

- 119805

### Tests

- `tests/openvino/native/quantization/test_graphs.py`
- `tests/openvino/native/test_model_transformer.py`

On top of https://github.com/openvinotoolkit/openvino/pull/21034 - **Merged**

### CODE REVIEW

**Comment by l-bat** [comment]
@KodiaqQ  Are you planning to add a new `mode` parameter to the documentation?

**Comment by KodiaqQ** [comment]
> @KodiaqQ Are you planning to add a new `mode` parameter to the documentation?

Good catch, thanks. I'll add it.
But first, we need to discuss this parameter first.

File: nncf/quantization/fake_quantize.py
Comment (alexsu52) [inline nncf/quantization/fake_quantize.py:265]: What do you think about using `enum` instead of `str`?

```
class DestinationType(Enum):
   HF8 = "HF8"
   BF8 = "BF8"
```

**Review by alexsu52 — COMMENTED** [review]
Some general comments:
- FBC and BC support FakeConvert operations?
- What test plan do you propose for FP8 quantization?

File: nncf/quantization/advanced_parameters.py
Comment (alexsu52) [reply nncf/quantization/advanced_parameters.py:146]: Let's discuss it offline.

**Comment by KodiaqQ** [comment]
> * FBC and BC support FakeConvert operations?

Support.
> * What test plan do you propose for FP8 quantization?

I've tested several models, see results here - 119805. 
If it's not enough, I can plan the validation, but it would take an additional week before we can merge PR.

File: nncf/quantization/fake_quantize.py
Comment (KodiaqQ) [reply nncf/quantization/fake_quantize.py]: Done as discussed.

File: tests/openvino/tools/calibrate.py
Comment (KodiaqQ) [reply tests/openvino/tools/calibrate.py]: Done.

**Comment by KodiaqQ** [comment]
Ready to review, @l-bat, @alexsu52, @AlexKoff88, @andreyanufr

File: nncf/openvino/graph/model_transformer.py
Comment (l-bat) [inline nncf/openvino/graph/model_transformer.py:20]: Should we move to the latest opset by default in a separate PR?

**Review by l-bat — APPROVED** [review]

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:211]: There was no request to support these parameters. Do we need it?

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:217]: The same answer as above.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:226]: We should change all internal parameters no matter were they set or not. In this case, we don't want to process `ConvertParameters`.

File: nncf/quantization/advanced_parameters.py
Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:211]: FP8 is represented not as the `*quantize` layer itself. Highly likely it will be presented as a usual `Convert` layer in the future.
In this case, naming `FP8QuantizationParameters` is weird and may confuse users since they would like to see some `*quantize` layers after using `*QuantizationParameters`.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:191]: I think that we shouldn't do it for INT8 quantization, since it would affect user experience. If the user provided the wrong configuration, we should say about it, and not change the parameters as we want.
In this case, the existence of the `_redefine_defaults_based_on_mode` is really weird. But since we do workarounds for FP8, there is no other way to change it rapidly.

File: nncf/quantization/advanced_parameters.py
Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:104]: It's not possible, since it requires renaming for the dependent methods. For example, `HWConfig.get_quantization_mode_from_config_value` should be also renamed to return `scheme` instead of the `mode`.
Is it in the scope of this PR?

Comment (alexsu52) [reply nncf/quantization/advanced_parameters.py:104]: I hope, yes.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:191]: Let's discuss it offline

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:211]: How long will it take to support this?

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:217]: How long will it take to support this?

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:223]: If you don't use this parameter in the algorithm. Let's remove it.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:226]: Do you have test that check `QuantizationMode` parameter?

File: nncf/quantization/advanced_parameters.py
Comment (alexsu52) [reply nncf/quantization/advanced_parameters.py:209]: I suggested to use None. This is correlate with `_redefine_defaults_based_on_mode` comment.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:223]: This parameter was introduced in your PR. May I ask, why did we need it?

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:211]: Few days.

File: nncf/quantization/advanced_parameters.py
Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:104]: Also, the NNCF configuration files will become invalid when changing `mode -> scheme`.

Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:104]: I've tried to do it in the whole NNCF. Not sure that pre-commit validation would be successful, but let's see.
Also, I think such a major change requires additional validation (at least e2e?) since it affects NNCF configuration files.
I would like to notify @vshampor about this huge change since it affects the whole NNCF.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py]: Good catch, thanks. Added tests to `tests/common/quantization/test_minmax.py`.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (alexsu52) [reply nncf/quantization/algorithms/min_max/algorithm.py:227]: Let's discuss it offline, because there are several decisions on the table.

File: nncf/quantization/advanced_parameters.py
Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:104]: Reverted commit with the global renaming `mode -> scheme`.

File: nncf/quantization/advanced_parameters.py
Comment (AlexKoff88) [reply nncf/quantization/advanced_parameters.py:211]: Just follow the hands 😄 
nncf.**quantize**, nncf.**Quantization**Mode, nncf.Advanced**Quantization**Parameters, activations_**quantization**_params

I am pretty much sure most of the users are not interested in what happens inside the model. But it would be better to be consistent on the API level.
BTW, can we just keep `QuantizationParameters` parameters?

File: nncf/quantization/advanced_parameters.py
Comment (alexsu52) [reply nncf/quantization/advanced_parameters.py]: It looks like `nncf.qunatize(...)` will contain two `QuantizationMode` which mean different things.

Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py]: Renamed to `QuantizationScheme` and `QuantizationMode`. Updated imports for `QuantizationScheme` to keep backward compatibility.

**Review by alexsu52 — COMMENTED** [review]
It looks like this PR requires additional validation. I would suggest to run ptq conformance test and DLB calibrate test on small model scope.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:832]: So, it's not confusing when we introduce `FP8QuantizationParameters` for `FakeConvert` layers. It's confusing when we create `FakeConvert` with the existing `create_quantizer_insertion_command` method.
Ok, I'll create a new command.

Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py:832]: Such renaming & new command creation would affect `_apply_quantizer_insertion_transformations` in the `ModelTransformer` as well (code duplication increase, for example).

**Comment by alexsu52** [comment]
The code `ov_quantized_model = nncf.quantize(ov_model, calibration_dataset, mode=nncf.QuantizationMode.FP8_E5M2)` produces the following warnings:
image

What do you think about printing only warning about the fact that FP8 is experimental option and raise runtime error if user specified parameters is not compatible with FP8 mode?

**Comment by KodiaqQ** [comment]
> The code `ov_quantized_model = nncf.quantize(ov_model, calibration_dataset, mode=nncf.QuantizationMode.FP8_E5M2)` produces the following warnings: image
> 
> What do you think about printing only warning about the fact that FP8 is experimental option and raise runtime error if user specified parameters is not compatible with FP8 mode?

Ok, I'll remove these warnings and add `RuntimeError` instead of the redefining provided options.

File: nncf/quantization/algorithms/min_max/algorithm.py
Comment (KodiaqQ) [reply nncf/quantization/algorithms/min_max/algorithm.py]: Updated.

**Comment by KodiaqQ** [comment]
@alexsu52, please, review.

File: nncf/quantization/algorithms/min_max/backend.py
Comment (alexsu52) [inline nncf/quantization/algorithms/min_max/backend.py:172]: typo

**Review by alexsu52 — COMMENTED** [review]
LGTM, @KodiaqQ please merge it after passing additional validation.

**Review by AlexKoff88 — APPROVED** [review]

**Comment by KodiaqQ** [comment]
> LGTM, @KodiaqQ please merge it after passing additional validation.

Conformance validation run - 229 (manual).
Observed errors were fixed with the latest commit (for per-channel activations aka depth-wise layers).
Is it enough, @alexsu52?

**Review by alexsu52 — APPROVED** [review]

**Pull request merged**
